opt: panic in execbuilder

```
SELECT subq_0.c0 AS c0 FROM (SELECT (SELECT rolcanlogin FROM pg_catalog.pg_roles LIMIT 1 OFFSET 2) AS c0, ref_0.attrs AS c1 FROM crdb_internal.kv_store_status AS ref_0 WHERE true) AS subq_0 WHERE (SELECT descriptor_id FROM crdb_internal.index_columns LIMIT 1 OFFSET 2) >= CASE WHEN CASE WHEN (SELECT umoptions FROM pg_catalog.pg_user_mapping LIMIT 1 OFFSET 6) = (SELECT umoptions FROM pg_catalog.pg_user_mapping LIMIT 1 OFFSET 14) THEN subq_0.c1 ELSE subq_0.c1 END IS NOT NULL THEN pg_catalog.extract(CAST(CASE WHEN ((CAST(NULL AS DECIMAL) <= CAST(NULL AS FLOAT8)) OR ((SELECT client_addr FROM pg_catalog.pg_stat_activity LIMIT 1 OFFSET 4) >= (SELECT client_addr FROM pg_catalog.pg_stat_activity LIMIT 1 OFFSET 5))) OR (EXISTS (SELECT ref_1.config_yaml AS c0, ref_1.config_yaml AS c1, subq_0.c0 AS c2, ref_1.config_yaml AS c3 FROM crdb_internal.zones AS ref_1 WHERE subq_0.c0 IS NOT NULL LIMIT 52)) THEN pg_catalog.version() ELSE pg_catalog.version() END AS TEXT), CAST(pg_catalog.current_date() AS DATE)) ELSE pg_catalog.extract(CAST(CASE WHEN ((CAST(NULL AS DECIMAL) <= CAST(NULL AS FLOAT8)) OR ((SELECT client_addr FROM pg_catalog.pg_stat_activity LIMIT 1 OFFSET 4) >= (SELECT client_addr FROM pg_catalog.pg_stat_activity LIMIT 1 OFFSET 5))) OR (EXISTS (SELECT ref_1.config_yaml AS c0, ref_1.config_yaml AS c1, subq_0.c0 AS c2, ref_1.config_yaml AS c3 FROM crdb_internal.zones AS ref_1 WHERE subq_0.c0 IS NOT NULL LIMIT 52)) THEN pg_catalog.version() ELSE pg_catalog.version() END AS TEXT), CAST(pg_catalog.current_date() AS DATE)) END LIMIT 107
```

```
panic: cannot map variable 28 to an indexed var [recovered]
        panic: panic while executing 1 statements: SELECT _._ AS _ FROM (SELECT (SELECT _ FROM _._ LIMIT _ OFFSET _) AS _, _._ AS _ FROM _._ AS _ WHERE _) AS _ WHERE (SELECT _ FROM _._ LIMIT _ OFFSET _) >= CASE WHEN CASE WHEN (SELECT _ FROM _._ LIMIT _ OFFSET _) = (SELECT _ FROM _._ LIMIT _ OFFSET _) THEN _._ ELSE _._ END IS NOT _ THEN extract(CAST(CASE WHEN ((CAST(_ AS DECIMAL) <= CAST(_ AS FLOAT8)) OR ((SELECT _ FROM _._ LIMIT _ OFFSET _) >= (SELECT _ FROM _._ LIMIT _ OFFSET _))) OR (EXISTS (SELECT _._ AS _, _._ AS _, _._ AS _, _._ AS _ FROM _ [...]; caused by cannot map variable 28 to an indexed var

goroutine 55 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc425372000, 0x68c4de0, 0xc420be6380, 0x5f54860, 0xc425e384c0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:675 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc425372000, 0x68c4de0, 0xc420be6380)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:414 +0x61
panic(0x5f54860, 0xc425e384c0)
        /usr/local/Cellar/go/1.10.2/libexec/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).indexedVar(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0xc40000001c, 0x0, 0xc428cae1b0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:109 +0x243
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildVariable(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x2800000037, 0x68b83e0, 0xc428cae1d8, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:101 +0x8b
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x2800000037, 0x2800000037, 0x0, 0x6153c40, 0xc425e3ee10)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildComparison(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x4a00000039, 0xc0, 0xc428cae2b8, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:177 +0xa1
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x4a00000039, 0x4a00000039, 0xef7cbb67c2154b1e, 0x401189d, 0xc428cae330)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildBoolean(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x330000003a, 0x4, 0xc428cae400, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:141 +0x155
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557c40, 0xc42572e780, 0x330000003a, 0x330000003a, 0x0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildSelect(0xc4265577c0, 0xc42572e780, 0x50000003b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x68dbba0, 0x7dc8f40)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:332 +0x226
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc4265577c0, 0xc42572e780, 0x50000003b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x9, 0xc42655e280)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:98 +0x98d
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildLimitOffset(0xc4265577c0, 0xc42572e780, 0x1f0000003f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x68dbba0, 0x7dc8f40)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:692 +0xe3
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc4265577c0, 0xc42572e780, 0x1f0000003f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7dc8f40, 0x68dbba0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:114 +0x10fa
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc4265577c0, 0xc42572e780, 0x1f0000003f, 0xc42572e780, 0x1f0000003f, 0x30, 0x61f1460)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:57 +0xb4
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildExistsSubquery(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3200000064, 0x68d91a0, 0xc428caefc0, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:374 +0x97
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3200000064, 0x3200000064, 0xc425e3b560, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildBoolean(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3500000065, 0x68d9260, 0xc428caf108, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:146 +0x287
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3500000065, 0x3500000065, 0x1, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCase(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6400000068, 0x7dc8f40, 0xc428caf260, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:250 +0x2ae
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6400000068, 0x6400000068, 0x20, 0x20, 0xc425e3d5c0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCast(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6300000069, 0x1, 0xc428caf328, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:274 +0x92
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6300000069, 0x6300000069, 0x2, 0x12, 0x68d8f60)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildFunction(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x670000006b, 0x68d91a0, 0xc428caf438, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:215 +0x148
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x670000006b, 0x670000006b, 0xc425e3b4d0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCase(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6400000077, 0x0, 0xc428caf590, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:254 +0x355
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x6400000077, 0x6400000077, 0x0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildComparison(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3a00000078, 0x60, 0xc428caf670, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:177 +0xa1
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3a00000078, 0x3a00000078, 0x5293892, 0x0, 0xc428caf6e8)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildBoolean(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3300000079, 0x2, 0xc428caf7b8, 0x52a025a, 0xc42572e780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:141 +0x155
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc4265577c0, 0xc426557940, 0xc42572e780, 0x3300000079, 0x3300000079, 0x0, 0x2000, 0x1000000000000)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:87 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildSelect(0xc4265577c0, 0xc42572e780, 0x50000007a, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc426561901, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:332 +0x226
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc4265577c0, 0xc42572e780, 0x50000007a, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc428cafda0, 0x53669b7)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:98 +0x98d
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildLimitOffset(0xc4265577c0, 0xc42572e780, 0x1f0000007e, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc4256908e0, 0x1)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:692 +0xe3
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc4265577c0, 0xc42572e780, 0x1f0000007e, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:114 +0x10fa
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildProject(0xc4265577c0, 0xc42572e780, 0x60000007f, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc428cb0500, 0x53669b7)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:348 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc4265577c0, 0xc42572e780, 0x60000007f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f, 0xc428cb0940)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:101 +0xa53
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc4265577c0, 0xc42572e780, 0x60000007f, 0x62d1360, 0xc428cb0970, 0x4015198, 0x40)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:57 +0xb4
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc4265577c0, 0xc4265577c0, 0xc426557740, 0xc42572e780, 0x60000007f)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:46 +0x51
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc425372438, 0x68c4ea0, 0xc428c2f620, 0x68c84a0, 0xc427dc5640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:405 +0x584
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optionallyUseOptimizer(0xc425372438, 0x68c4ea0, 0xc428c2f620, 0xc4204421e0, 0xd, 0xc42539001b, 0x9, 0x0, 0x1, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/planner.go:534 +0xdb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc425372000, 0x68c4ea0, 0xc428c2f620, 0x68c84a0, 0xc427dc5640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:751 +0x192
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc425372000, 0x68c4ea0, 0xc428c2f620, 0x68c84a0, 0xc427dc5640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:401 +0xb82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc425372000, 0x68c4ea0, 0xc428c2f620, 0x68c84a0, 0xc427dc5640, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc425372000, 0x68c4de0, 0xc420be6380, 0xc4204001b0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1079 +0x2186
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc42054e120, 0x68c4de0, 0xc420be6380, 0xc42539001b, 0x9, 0xc42539000d, 0x4, 0x0, 0x0, 0x68a6d60, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x1bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc42054e120, 0x68c4de0, 0xc420be6380, 0xc425418000, 0x5400, 0x15000, 0xc420d4e180, 0xc4204001b0, 0xc4204001a0, 0xc42042c070, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:267 +0x122
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:266 +0xf04
```

--- a/pkg/sql/opt/exec/execbuilder/scalar_builder.go
+++ b/pkg/sql/opt/exec/execbuilder/scalar_builder.go
@@ -335,9 +335,21 @@ func (b *Builder) buildAnyScalar(ctx *buildScalarCtx, ev memo.ExprView) (tree.Ty
 	return tree.NewTypedComparisonExprWithSubOp(tree.Any, cmp, left, right), nil
 }
 
+func (b *Builder) buildUnsupportedExpr(
+	ctx *buildScalarCtx, ev memo.ExprView,
+) (tree.TypedExpr, error) {
+	return ev.Private().(tree.TypedExpr), nil
+}
+
 func (b *Builder) buildAny(ctx *buildScalarCtx, ev memo.ExprView) (tree.TypedExpr, error) {
+	input := ev.Child(0)
+	// We cannot execute correlated subqueries.
+	if !input.Logical().Relational.OuterCols.Empty() {
+		return nil, b.decorrelationError()
+	}
+
 	// Build the execution plan for the input subquery.
-	plan, err := b.buildRelational(ev.Child(0))
+	plan, err := b.buildRelational(input)
 	if err != nil {
 		return nil, err
 	}
@@ -348,7 +360,7 @@ func (b *Builder) buildAny(ctx *buildScalarCtx, ev memo.ExprView) (tree.TypedExp
 		types.Types[val] = ev.Metadata().ColumnType(opt.ColumnID(key))
 	})
 
-	input := b.addSubquery(exec.SubqueryAnyRows, types, plan.root)
+	subqueryExpr := b.addSubquery(exec.SubqueryAnyRows, types, plan.root)
 
 	// Build the scalar value that is compared against each row.
 	scalar, err := b.buildScalar(ctx, ev.Child(1))
@@ -357,21 +369,21 @@ func (b *Builder) buildAny(ctx *buildScalarCtx, ev memo.ExprView) (tree.TypedExp
 	}
 
 	cmp := opt.ComparisonOpReverseMap[ev.Private().(opt.Operator)]
-	return tree.NewTypedComparisonExprWithSubOp(tree.Any, cmp, scalar, input), nil
-}
-
-func (b *Builder) buildUnsupportedExpr(
-	ctx *buildScalarCtx, ev memo.ExprView,
-) (tree.TypedExpr, error) {
-	return ev.Private().(tree.TypedExpr), nil
+	return tree.NewTypedComparisonExprWithSubOp(tree.Any, cmp, scalar, subqueryExpr), nil
 }
 
 func (b *Builder) buildExistsSubquery(
 	ctx *buildScalarCtx, ev memo.ExprView,
 ) (tree.TypedExpr, error) {
+	input := ev.Child(0)
+	// We cannot execute correlated subqueries.
+	if !input.Logical().Relational.OuterCols.Empty() {
+		return nil, b.decorrelationError()
+	}
+
 	// Build the execution plan for the subquery. Note that the subquery could
 	// have subqueries of its own which are added to b.subqueries.
-	root, err := b.build(ev.Child(0))
+	root, err := b.build(input)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/sql/opt/exec/execbuilder/testdata/subquery
+++ b/pkg/sql/opt/exec/execbuilder/testdata/subquery
@@ -219,9 +219,6 @@ join       ·      ·            (x, z)  ·
 ·          table  a@primary    ·       ·
 ·          spans  ALL          ·       ·
 
-query error could not decorrelate subquery
-SELECT * FROM abc WHERE EXISTS(SELECT * FROM (VALUES (a), (b)) WHERE column1=a)
-
 query TTTTT
 EXPLAIN (VERBOSE) SELECT ARRAY(SELECT x FROM b)
 ----
@@ -239,3 +236,64 @@ root                 ·            ·                       ("array")  ·
            └── scan  ·            ·                       (x)        ·
 ·                    table        b@primary               ·          ·
 ·                    spans        ALL                     ·          ·
+
+# Case where the plan has an apply join.
+query error could not decorrelate subquery
+SELECT * FROM abc WHERE EXISTS(SELECT * FROM (VALUES (a), (b)) WHERE column1=a)
+
+# Case where the EXISTS subquery still has outer columns in the subquery
+# (regression test for #28816).
+query error could not decorrelate subquery
+SELECT
+  subq_0.c1 AS c1
+FROM
+  (SELECT ref_0.attrs AS c1 FROM crdb_internal.kv_store_status AS ref_0) AS subq_0
+WHERE
+  5 >= CASE WHEN subq_0.c1 IS NOT NULL
+    THEN 5
+    ELSE pg_catalog.extract(
+      CAST(
+        CASE WHEN (
+          EXISTS(
+            SELECT ref_1.config_yaml AS c0
+            FROM crdb_internal.zones AS ref_1
+            WHERE subq_0.c1 IS NOT NULL
+          )
+        )
+        THEN pg_catalog.version()
+        ELSE pg_catalog.version()
+        END
+          AS STRING
+      ),
+      CAST(pg_catalog.current_date() AS DATE)
+    )
+    END
+
+# Case where the ANY subquery still has outer columns.
+query error could not decorrelate subquery
+SELECT
+  subq_0.c1 AS c1
+FROM
+  (SELECT ref_0.attrs AS c1 FROM crdb_internal.kv_store_status AS ref_0) AS subq_0
+WHERE
+  5 >= CASE WHEN subq_0.c1 IS NOT NULL
+    THEN 5
+    ELSE pg_catalog.extract(
+      CAST(
+        CASE
+        WHEN (
+          '12'::BYTES
+          = ANY (
+              SELECT ref_1.config_yaml AS c0
+              FROM crdb_internal.zones AS ref_1
+              WHERE subq_0.c1 IS NOT NULL
+            )
+        )
+        THEN pg_catalog.version()
+        ELSE pg_catalog.version()
+        END
+          AS STRING
+      ),
+      CAST(pg_catalog.current_date() AS DATE)
+    )
+    END

